src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,906 @@
+import java.text.MessageFormat;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Random;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Level;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+import mssql.security.provider.MD4;
+
+
+/**
+ * Provides an implementation of NTLM authentication
+ * 
+ * @see <a
+ *      href=https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-nlmp/b38c36ed-2804-4868-a9ff-8dd3182128e4">MS-NLMP:NTLM
+ *      Authentication Protocol</a>
+ */
+final class NTLMAuthentication extends SSPIAuthentication {
+    private final java.util.logging.Logger logger = java.util.logging.Logger
+            .getLogger("com.microsoft.sqlserver.jdbc.internals.NTLMAuthentication");
+
+    /**
+     * Section 2.2.1.1 NEGOTIATE_MESSAGE
+     *
+     * Section 2.2.1.2 CHALLENGE_MESSAGE
+     *
+     * Section 2.2.1.3 AUTHENCIATE_MESSAGE
+     *
+     * NTLM signature for all 3 messages - "NTLMSSP\0"
+     */
+    private static final byte[] NTLM_HEADER_SIGNATURE = {0x4e, 0x54, 0x4c, 0x4d, 0x53, 0x53, 0x50, 0x00};
+
+    // NTLM messages type
+    private static final int NTLM_MESSAGE_TYPE_NEGOTIATE = 0x00000001;
+    private static final int NTLM_MESSAGE_TYPE_CHALLENGE = 0x00000002;
+    private static final int NTLM_MESSAGE_TYPE_AUTHENTICATE = 0x00000003;
+
+    /**
+     * Section 2.2.2.7 NTLM v2: NTLMv2_CLIENT_CHALLENGE
+     * 
+     * <pre>
+     * RespType   current version(aka Responseversion) of the challenge response type. This field MUST be 0x01
+     * HiRespType max supported version (aka HiResponserversion) of the client response type.
+     * </pre>
+     */
+    private static final byte[] NTLM_CLIENT_CHALLENGE_RESPONSE_TYPE = {0x01, 0x01};
+
+    /**
+     * Section 2.2.2.7 NTLM v2: NTLMv2_CLIENT_CHALLENGE
+     *
+     * <pre>
+     * Reserved1  SHOULD be 0x0000
+     * Reserved2  SHOULD be 0x00000000
+     * Reserved3  SHOULD be 0x00000000
+     * </pre>
+     */
+    private static final byte[] NTLM_CLIENT_CHALLENGE_RESERVED1 = {0x00, 0x00};
+    private static final byte[] NTLM_CLIENT_CHALLENGE_RESERVED2 = {0x00, 0x00, 0x00, 0x00};
+    private static final byte[] NTLM_CLIENT_CHALLENGE_RESERVED3 = {0x00, 0x00, 0x00, 0x00};
+
+    /**
+     * Section 3.1.5.1.1 Client Initiates the NEGOTIATE_MESSAGE
+     * 
+     * If NTLM v2 authentication is used and the CHALLENGE_MESSAGE TargetInfo field (section 2.2.1.2) has an
+     * MsvAvTimestamp present, the client SHOULD NOT send the LmChallengeResponse and SHOULD send Z(24) instead.
+     *
+     * Section 2.2.2.1 AV_PAIR
+     *
+     * MsvAvTimestamp is always sent in the CHALLENGE_MESSAGE.
+     *
+     * Section 7 Appendix B: Product Behavior
+     *
+     * MsvAvTimestamp AV_PAIR type is not supported in Windows NT, Windows 2000, Windows XP, and Windows Server 2003.
+     */
+    private static final byte[] NTLM_LMCHALLENAGERESPONSE = {0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,

```

Comment:
`NTLM_LMCHALLENAGERESPONSE` is not used anywhere. Only length of it is used in `generateNtlmAuthenticate()`, which can be part of `NTLM_AUTHENTICATE_PAYLOAD_OFFSET`.

Reply:
it's used in authenticate message. It's nicer to separate so it's clear what it is instead of just lumping it into the payload?